neutrons/SNAPRed
# For EWM 1842 - UI integration normalization calibration (backend and frontend)

## Description of work
This story is to implement the UI workflow for a user to be able to execute a `Normalization Calibration` onto a `Diffraction Calibration` run. Along side this, the backend methods (e.i. Algorithm, recipe, service, etc.) have been added to perform the logical execution.

The workflow for `Normalization Calibration` is the following:
1. **NormalizationCalibrationRequestView.py & _triggerNormalizationCalibration()**
Contains three empty input text fields for run number, background run number, and smoothing parameter
Contains two drop down menus for selecting samples and grouping definition
All values are then updated within other views so that user inputs are viewable in all steps and values remain consistent for the final saving step
callNormalizationCalibration (self, groupingIndex, smoothingParameter) -> Call is made to CalibrationService.py @ "calibration/normalization" where all required ingredients are derived and orchestrated for the recipe

2. **CalibrationNormalizationRecipe() -> (normalizationIngredients, groceries)**
Within CalibrationNormalizationRecipe, ingredients are chopped and groceries are fetched to initialize all the required workspaces for the main backend logic -> CalibrationNormalizationAlgo()
Properties are set and a try / except method is used to execute the algo
This returns two workspaces [focusWorkspace, smoothedWorkspace]

3. **CalibrationNormalizationAlgo()**
Inputs for Algo:
InputWorkspace
BackgroundWorkspace
Ingredients
Output for Algo:
FocusWorkspace
SmoothWorkspace
Execution logic of the Algo
All input ingredients are chopped and orchestrated
RawVanadiumCorrectionAlgorithm (inputWS, backgroundWS, reductionIngredients, CalibrantSample, outputWS) **NOTE:** A current input is a calibration workspace containing the diffc table, this should be removed from within [Story 2989](https://github.com/ccm/resource/itemName/com.ibm.team.workitem.WorkItem/2989)) [Therefore, it has been removed already.] -> This outputs a rawVanadium (absorption corrected WS)
LoadGroupingDefinition (GroupingFilename, InstrumentDonor=rawVanadium(should be lite), outputWS)
ApplyDiffcal(inputWS, calibrationWS)
ConvertUnits(InputWS, Emode="Elastic", Target="dSpacing", outputWS, ConvertFromPointData=True)
DiffractionFocussing(inputWS, groupingWS, outputWS, PreserveEvents=True)
RebinRagged(inputWS, XMin, XMax, Delta, outputWS, PreserveEvents=True) **NOTE:** This is done for all pixel grouping parameters
CloneWorkspace(inputWS, outputWS)
SmoothDataExcludingPeaks(inputWS, ingredients, outputWS)
WashDishes(inputWS, backgroundWS, rawVanadiumWS, groupingWS, focusedWS, clonedWS)
Return (focusedWS, smoothWS)

4. **_SpecifyNormalizationView & _specifyNormalization()**
This view incorporates all the original user input and these user inputs are not editable, they are present just to view however there are two inputs that can be modifiable during this step:
smoothing parameter (0 to 1)
grouping schema -> All, Column, Bank
number of graphed spectra
This view also includes a two row graph matrix. The size or number of graphs within each row will be dependent on the type of grouping schema selected by the user ->
If All: numGraphs = 1 (therefore two rows of 1 graph each [2 graphs total])
If Bank: numGraphs = 2 (therefore two rows of 2 graphs each [4 graphs total])
If Column: numGraphs= 6 (therefore two rows 6 graphs each [12 graphs total])
As the user updates the smoothing parameter on the slider, or the selected grouping schema, a signal is emitted to pass on the new grouping schema index or smoothing parameter value
The workflow looks like this:
changedGrouping, changedSmoothingParameter -> send a signal with new values to the workflow and pass them as inputs to the "calibration/normalization" backend service call passing new parameters (everything else remains the same)
this then returns two workspaces after some computation time -> sends a signal with the new workspaces to the view to a update workspaces method that updates the figure with new workspace values
rinse and repeat till the user finds their desired configuration for smoothing parameter and grouping schema -> clean up all workspaces other than the final ones -> pass all configuration data and workspaces to saving step

5.   **_SpecifyNormalizationCalibrationView & _saveNormalizationCalibration()**
All the previous data is fed into this step. The user is able to see their final configurations within this view before executing the saving to disk.
None of the inputs are editable
The data from the previous steps are orchestrated into a NormalizationRecord (similar to a CalibrationRecord) and saved within the versioned Calibration directory.



## Explanation of work



## To test

This workflow should be tested through launching the GUI. Once the GUI is launched from within analysis, there will be a new tab that appears called **Normalization Calibration**. Within this tab the tester should put in inputs. Currently, I've been testing with the following:

runNumber = 58810
background = 58813
Smoothing Parameter = 0.5
Sample = Silicon_NIST_640D_001.json (the only sample that will work for now due to defect mentioned before)
Grouping = Column

Then hit continue and the workflow should commence. The execution progress will be viewable within the launcher window of SNAPRed. This may take a minute to execute. Once on the "Tweak Parameters" tab, the user can modify the smoothing parameter through the slider or the text box and the grouping file. 

**NOTE:** In order for the re-execution with new parameters to proceed, the user must click into the text box containing the smoothing parameter and hit **ENTER** to proceed.

### For CIS Testing

The CIS will follow the same steps listed above but will also ensure that the save methods are executed correctly. Since we don't have write permission and we haven't implemented mocking this yet in testing, please make sure that the files are saved within the correct directory `/SNS/SNAP/shared/Calibration/Powder/ (state sha) / (some version number) /`.

Once the tester is satisfied with their selection, they can hit "Continue" to finish the workflow.



## Link to EWM item


[EWM #3232](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20%28Change%20Management%29#action=com.ibm.team.workitem.viewWorkItem&id=3232)
